File: src/ios/TikTokBusinessSdkPlugin.m
#import "TikTokBusinessSdkPlugin.h"
#import <TikTokBusinessSDK/TikTokBusinessSDK.h>

@implementation TikTokBusinessSdkPlugin

/**
 * Get platform version
 */
- (void)getPlatformVersion:(CDVInvokedUrlCommand*)command {
    NSString* version = [NSString stringWithFormat:@"iOS %@", [[UIDevice currentDevice] systemVersion]];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:version];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Initialize TikTok Business SDK
 */
- (void)initTiktokBusinessSdk:(CDVInvokedUrlCommand*)command {
    NSDictionary* config = [command.arguments objectAtIndex:0];
    
    NSString* accessToken = [config objectForKey:@"accessToken"];
    NSString* appId = [config objectForKey:@"appId"];
    NSString* ttAppId = [config objectForKey:@"ttAppId"];
    BOOL openDebug = [[config objectForKey:@"openDebug"] boolValue];
    
    // enableAutoIapTrack defaults to true
    // If not provided or explicitly set to false, disable payment tracking
    id enableAutoIapTrackValue = [config objectForKey:@"enableAutoIapTrack"];
    BOOL enableAutoIapTrack = (enableAutoIapTrackValue == nil) ? YES : [enableAutoIapTrackValue boolValue];
    
    // Note: disableAutoEnhancedDataPostbackEvents is not available in iOS SDK
    // BOOL disableAutoEnhancedDataPostbackEvents = [[config objectForKey:@"disableAutoEnhancedDataPostbackEvents"] boolValue];
    
    if (!accessToken || [accessToken length] == 0 ||
        !appId || [appId length] == 0 ||
        !ttAppId || [ttAppId length] == 0) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                                          messageAsString:@"accessToken, appId, and ttAppId are required"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    
    // TikTok SDK configWithAccessToken method expects NSString for tiktokAppId
    // (despite the property being NSNumber in the header)
    TikTokConfig* ttConfig = [TikTokConfig configWithAccessToken:accessToken 
                                                           appId:appId 
                                                     tiktokAppId:ttAppId];
    
    if (openDebug) {
        [ttConfig enableDebugMode];
    }
    
    if (!enableAutoIapTrack) {
        [ttConfig disablePaymentTracking];
    }
    
    [ttConfig setLogLevel:TikTokLogLevelDebug];
    
    [TikTokBusiness initializeSdk:ttConfig completionHandler:^(BOOL success, NSError * _Nullable error) {
        CDVPluginResult* pluginResult;
        
        if (success) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        } else {
            NSString* errorMessage = error ? error.localizedDescription : @"Initialization failed";
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                              messageAsString:errorMessage];
        }
        
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

/**
 * Set user identity
 */
- (void)setIdentify:(CDVInvokedUrlCommand*)command {
    NSDictionary* identity = [command.arguments objectAtIndex:0];
    
    NSString* externalId = [identity objectForKey:@"externalId"];
    NSString* externalUserName = [identity objectForKey:@"externalUserName"];
    NSString* phoneNumber = [identity objectForKey:@"phoneNumber"];
    NSString* email = [identity objectForKey:@"email"];
    
    if (!externalId || [externalId length] == 0) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                                          messageAsString:@"externalId is required"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    
    [TikTokBusiness identifyWithExternalID:externalId 
                          externalUserName:externalUserName ?: @"" 
                               phoneNumber:phoneNumber ?: @"" 
                                     email:email ?: @""];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Track TikTok event
 */
- (void)trackTTEvent:(CDVInvokedUrlCommand*)command {
    NSDictionary* eventData = [command.arguments objectAtIndex:0];
    
    NSString* eventName = [eventData objectForKey:@"eventName"];
    NSString* eventId = [eventData objectForKey:@"eventId"];
    
    if (!eventName || [eventName length] == 0) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                                          messageAsString:@"eventName is required"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    
    [TikTokBusiness trackEvent:eventName withId:eventId ?: @""];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Track TikTok event with custom properties
 */
- (void)trackTTEventWithCustomData:(CDVInvokedUrlCommand*)command {
    NSDictionary* eventData = [command.arguments objectAtIndex:0];
    
    NSString* eventName = [eventData objectForKey:@"eventName"];
    NSString* eventId = [eventData objectForKey:@"eventId"];
    NSDictionary* properties = [eventData objectForKey:@"properties"];
    
    if (!eventName || [eventName length] == 0) {
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR 
                                                          messageAsString:@"eventName is required"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    
    // Create TikTokBaseEvent with custom properties
    TikTokBaseEvent* customEvent = [[TikTokBaseEvent alloc] initWithEventName:eventName];
    
    if (eventId && [eventId length] > 0) {
        [customEvent addPropertyWithKey:@"event_id" value:eventId];
    }
    
    // Add custom properties
    if (properties && [properties isKindOfClass:[NSDictionary class]]) {
        for (NSString* key in properties) {
            id value = properties[key];
            [customEvent addPropertyWithKey:key value:value];
        }
    }
    
    [TikTokBusiness trackTTEvent:customEvent];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

/**
 * Logout current user
 */
- (void)logout:(CDVInvokedUrlCommand*)command {
    [TikTokBusiness logout];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

@end

File: src/android/TikTokBusinessSdkPlugin.java
package com.tiktok.cordova;

import android.content.Context;
import com.tiktok.TikTokBusinessSdk;
import com.tiktok.appevents.base.EventName;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

/**
 * TikTok Business SDK Cordova Plugin for Android
 */
public class TikTokBusinessSdkPlugin extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        Context context = this.cordova.getActivity().getApplicationContext();

        switch (action) {
            case "getPlatformVersion":
                this.getPlatformVersion(callbackContext);
                return true;

            case "initTiktokBusinessSdk":
                JSONObject config = args.getJSONObject(0);
                this.initTiktokBusinessSdk(config, callbackContext, context);
                return true;

            case "setIdentify":
                JSONObject identity = args.getJSONObject(0);
                this.setIdentify(identity, callbackContext);
                return true;

            case "trackTTEvent":
                JSONObject eventData = args.getJSONObject(0);
                this.trackTTEvent(eventData, callbackContext);
                return true;

            case "trackTTEventWithCustomData":
                JSONObject customEventData = args.getJSONObject(0);
                this.trackTTEventWithCustomData(customEventData, callbackContext);
                return true;

            case "logout":
                this.logout(callbackContext);
                return true;

            default:
                return false;
        }
    }

    /**
     * Get platform version
     */
    private void getPlatformVersion(CallbackContext callbackContext) {
        String version = "Android " + android.os.Build.VERSION.RELEASE;
        callbackContext.success(version);
    }

    /**
     * Initialize TikTok Business SDK
     */
    private void initTiktokBusinessSdk(JSONObject config, CallbackContext callbackContext, Context context) {
        try {
            String accessToken = config.getString("accessToken");
            String appId = config.getString("appId");
            String ttAppId = config.getString("ttAppId");
            boolean openDebug = config.optBoolean("openDebug", false);
            boolean enableAutoIapTrack = config.optBoolean("enableAutoIapTrack", true);
            boolean disableAutoEnhancedDataPostbackEvents = config.optBoolean("disableAutoEnhancedDataPostbackEvents", false);

            if (accessToken == null || accessToken.isEmpty() ||
                appId == null || appId.isEmpty() ||
                ttAppId == null || ttAppId.isEmpty()) {
                callbackContext.error("accessToken, appId, and ttAppId are required");
                return;
            }

            TikTokBusinessSdk.TTConfig ttConfig = new TikTokBusinessSdk.TTConfig(context, accessToken);
            ttConfig.setAppId(appId);
            ttConfig.setTTAppId(ttAppId);

            if (openDebug) {
                ttConfig.openDebugMode();
            }

            if (enableAutoIapTrack) {
                ttConfig.enableAutoIapTrack();
            }

            if (disableAutoEnhancedDataPostbackEvents) {
                ttConfig.disableAutoEnhancedDataPostbackEvent();
            }

            TikTokBusinessSdk.initializeSdk(ttConfig, new TikTokBusinessSdk.TTInitCallback() {
                @Override
                public void success() {
                    callbackContext.success();
                }

                @Override
                public void fail(int code, String msg) {
                    callbackContext.error("Initialization failed: " + msg + " (code: " + code + ")");
                }
            });

        } catch (JSONException e) {
            callbackContext.error("Invalid configuration: " + e.getMessage());
        }
    }

    /**
     * Set user identity
     */
    private void setIdentify(JSONObject identity, CallbackContext callbackContext) {
        try {
            String externalId = identity.getString("externalId");
            String externalUserName = identity.optString("externalUserName", null);
            String phoneNumber = identity.optString("phoneNumber", null);
            String email = identity.optString("email", null);

            if (externalId == null || externalId.isEmpty()) {
                callbackContext.error("externalId is required");
                return;
            }

            TikTokBusinessSdk.identify(externalId, externalUserName, phoneNumber, email);
            callbackContext.success();

        } catch (JSONException e) {
            callbackContext.error("Invalid identity data: " + e.getMessage());
        }
    }

    /**
     * Track TikTok event
     */
    private void trackTTEvent(JSONObject eventData, CallbackContext callbackContext) {
        try {
            String eventName = eventData.getString("eventName");
            String eventId = eventData.optString("eventId", null);

            if (eventName == null || eventName.isEmpty()) {
                callbackContext.error("eventName is required");
                return;
            }

            // Try to use predefined EventName enum first, fallback to custom event string
            try {
                EventName event = EventName.valueOf(eventName);
                
                if (eventId != null && !eventId.isEmpty()) {
                    TikTokBusinessSdk.trackTTEvent(event, eventId);
                } else {
                    TikTokBusinessSdk.trackTTEvent(event);
                }
            } catch (IllegalArgumentException e) {
                // Not a predefined event, use custom event with string
                if (eventId != null && !eventId.isEmpty()) {
                    TikTokBusinessSdk.trackEvent(eventName, eventId);
                } else {
                    TikTokBusinessSdk.trackEvent(eventName);
                }
            }

            callbackContext.success("Event tracked successfully");

        } catch (JSONException e) {
            callbackContext.error("Invalid event data: " + e.getMessage());
        } catch (Exception e) {
            callbackContext.error("Track event failed: " + e.getMessage());
        }
    }

    /**
     * Track TikTok event with custom properties
     */
    private void trackTTEventWithCustomData(JSONObject eventData, CallbackContext callbackContext) {
        try {
            String eventName = eventData.getString("eventName");
            String eventId = eventData.optString("eventId", null);
            JSONObject properties = eventData.optJSONObject("properties");

            if (eventName == null || eventName.isEmpty()) {
                callbackContext.error("eventName is required");
                return;
            }

            // Create properties object with custom properties
            JSONObject props = new JSONObject();
            
            if (eventId != null && !eventId.isEmpty()) {
                props.put("event_id", eventId);
            }

            // Add custom properties
            if (properties != null) {
                Iterator<String> keys = properties.keys();
                while (keys.hasNext()) {
                    String key = keys.next();
                    Object value = properties.get(key);
                    props.put(key, value);
                }
            }

            TikTokBusinessSdk.trackEvent(eventName, props);
            callbackContext.success("Event with custom data tracked successfully");

        } catch (JSONException e) {
            callbackContext.error("Invalid event data: " + e.getMessage());
        } catch (Exception e) {
            callbackContext.error("Track event with custom data failed: " + e.getMessage());
        }
    }

    /**
     * Logout current user
     */
    private void logout(CallbackContext callbackContext) {
        TikTokBusinessSdk.logout();
        callbackContext.success();
    }
}
